File: client/src/main/scala/scalive/AgentLoader.scala
package scalive

// http://docs.oracle.com/javase/6/docs/jdk/api/attach/spec/index.html
import com.sun.tools.attach.VirtualMachine

import java.io.File
import java.net.URLClassLoader

object AgentLoader {
 /**
   * @param args <jarpath> [class loader id]
   *
   * jarpath is the absolute path this directory:
   *
   * {{{
   * jarpath/
   *   scalive-agent.jar
   *   scalive-client.jar
   *   scalive-repl.jar
   *
   *   2.10.3/
   *     scala-library.jar
   *     scala-compiler.jar
   *     scala-reflect.jar
   *
   *   [Other Scala versions]
   * }}}
   */
  def main(args: Array[String]) {
    if (args.length != 1 && args.length != 2 && args.length != 3) {
      println("Arguments: <jarpath> [pid] [class loader id]")
      return
    }

    addToolsDotJarToClasspath()

    if (args.length == 1) {
      listJvmProcesses()
      return
    }

    val jarpath = args(0)
    val pid     = args(1)
    val clId    = if (args.length == 3) args(2) else ""
    loadAgent(jarpath, pid, clId)
  }

  /**
   * com.sun.tools.attach.VirtualMachine is in tools.jar, which is not in
   * classpath by default:
   *
   * {{{
   * jdk/
   *   bin/
   *     java
   *     javac
   *   jre/
   *     java
   *   lib/
   *     tools.jar
   * }}}
   */
  private def addToolsDotJarToClasspath() {
    val path = System.getProperty("java.home") + "/../lib/tools.jar"
    Classpath.addPath(ClassLoader.getSystemClassLoader.asInstanceOf[URLClassLoader], path)
  }

  private def listJvmProcesses() {
    println("JVM processes:")
    println("#pid\tDisplay name")

    val it = VirtualMachine.list.iterator()
    while (it.hasNext()) {
      val vmd = it.next()
      println(vmd.id + "\t" + vmd.displayName)
    }
  }

  private def loadAgent(jarpath: String, pid: String, clId: String) {
    val agentJar = Classpath.findJar(jarpath, "scalive-agent")
    val vm       = VirtualMachine.attach(pid)
    val port     = Client.getFreePort()
    vm.loadAgent(agentJar, jarpath + " " + port + " " + clId)

    Runtime.getRuntime.addShutdownHook(new Thread {
      override def run() { vm.detach() }
    })

    Client.connectToRepl(port)
  }
}

File: src/main/java/scalive/AgentLoader.java
package scalive;

// http://docs.oracle.com/javase/6/docs/jdk/api/attach/spec/index.html
import com.sun.tools.attach.VirtualMachine;
import com.sun.tools.attach.VirtualMachineDescriptor;

import java.io.File;
import java.net.URLClassLoader;
import java.util.Iterator;

public class AgentLoader {
    /**
     * @param args <jarSearchDir1>[<File.pathSeparator><jarSearchDir2>...] [pid]
     *
     * <pre>{@code
     * jarSearchDir/
     *   scalive.jar
     *
     *   scala-library-2.11.0.jar
     *   scala-compiler-2.11.0.jar
     *   scala-reflect-2.11.0.jar
     *
     *   [Other Scala versions]
     * }</pre>
     */
    public static void main(String[] args) throws Exception {
        if (args.length != 1 && args.length != 2) {
            System.out.println("Arguments: <jarSearchDir1>[<File.pathSeparator><jarSearchDir2>...] [pid]");
            return;
        }

        addToolsDotJarToClasspath();

        if (args.length == 1) {
            listJvmProcesses();
            return;
        }

        String jarSearchDirs = args[0];
        String pid      = args[1];
        loadAgent(jarSearchDirs, pid);
    }

    /**
     * com.sun.tools.attach.VirtualMachine is in tools.jar, which is not in
     * classpath by default:
     *
     * <pre>{@code
     * jdk/
     *   bin/
     *     java
     *     javac
     *   jre/
     *     java
     *   lib/
     *     tools.jar
     * }</pre>
     */
    private static void addToolsDotJarToClasspath() throws Exception {
        String path = System.getProperty("java.home") + "/../lib/tools.jar";
        Classpath.addPath((URLClassLoader) ClassLoader.getSystemClassLoader(), path);
    }

    private static void listJvmProcesses() {
        System.out.println("JVM processes:");
        System.out.println("#pid\tDisplay name");

        Iterator<VirtualMachineDescriptor> it = VirtualMachine.list().iterator();
        while (it.hasNext()) {
            VirtualMachineDescriptor vmd = it.next();
            System.out.println(vmd.id() + "\t" + vmd.displayName());
        }
    }

    private static void loadAgent(String jarSearchDirs, String pid) throws Exception {
        final String[]       ary      = jarSearchDirs.split(File.pathSeparator);
        final String         agentJar = Classpath.findJar(ary, "scalive");
        final VirtualMachine vm       = VirtualMachine.attach(pid);
        final int            port     = Client.getFreePort();

        vm.loadAgent(agentJar, jarSearchDirs + " " + port);
        Runtime.getRuntime().addShutdownHook(new Thread() {
            @Override public void run() {
                try {
                    vm.detach();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });

        Client.connectToRepl(port);
    }
}
